### PR TITLE
Textcnn

### PR DESCRIPTION
**Pull request by lhz1029**
A first pass.

I'm not quite sure how to best refactor the tests now that there are many base models and many targets but not every pair is possible. I've just been manually changing the config for now to test out textcnn, but ideally this is part of the testing suite.

### CODE REVIEW

File: finetune/base_models/textcnn/featurizer.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            initializer=tf.random_normal_initializer(stddev=config.weight_stddev)
+        )
+        if config.train_embeddings:
+            embed_weights = dropout(embed_weights, config.embed_p_drop, train)
+        else:
+            embed_weights = tf.stop_gradient(embed_weights)
+
+        X = tf.reshape(X, [-1, config.max_length, 2])
+
+        h = embed(X, embed_weights)
```
Comment (benleetownsend) [inline finetune/base_models/textcnn/featurizer.py]: This function includes positional embeddings. This is *probably* but not definitely undesirable. You can easily circumvent this by passing `X[:, :, :1,]` This is because the format of X is [batch_size, sequence_len, 2] where the 2 dim is 
(word embeddings, positional embeddings)

File: finetune/base_models/textcnn/featurizer.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                padding='same',
+                activation=tf.nn.relu,
+                name='conv' + str(i)
+            )
+            conv_layers.append(conv)
+            pool = tf.reduce_max(conv, axis=1)
+            pool_layers.append(pool)
+
+        # Concat the output of the convolutional layers for use in sequence embedding
+        conv_seq = tf.concat(conv_layers, axis=2)
```
Comment (benleetownsend) [inline finetune/base_models/textcnn/featurizer.py]: Love it.

File: finetune/base_models/textcnn/featurizer.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            conv = tf.layers.conv1d(
+                inputs=h,
+                filters=config.num_filters_per_size,
+                kernel_size=kernel_size,
+                padding='same',
+                activation=tf.nn.relu,
+                name='conv' + str(i)
+            )
+            conv_layers.append(conv)
+            pool = tf.reduce_max(conv, axis=1)
```
Comment (benleetownsend) [inline finetune/base_models/textcnn/featurizer.py]: Just realised you should probably mask the padding tokens before max pooling

this  could potentially look something like

```
        mask = tf.expand_dims(1.0 - tf.sequence_mask(pool_idx, maxlen=tf.shape(conv)[1], dtype=tf.float32), -1)
        max_pooled = tf.reduce_max(conv + mask * -1e9, 1)
```

Comment (benleetownsend) [reply finetune/base_models/textcnn/featurizer.py]: Actually, you will probably also want to mask the 0th token which is a BOS/start token.

File: finetune/base_models/textcnn/model.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+from finetune.base_models.textcnn.featurizer import textcnn_featurizer
+
+class TextCNNModel(SourceModel):
+
+    encoder = GPTEncoder
+    featurizer = textcnn_featurizer
+    kernel_sizes = [2, 3, 4]
+    settings = {
+        'n_embed_featurizer': 768,
+        'n_heads': 12,
```
Comment (benleetownsend) [inline finetune/base_models/textcnn/model.py]: this transformer related stuff can go now.

File: finetune/base_models/textcnn/model.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+from finetune.base_models.gpt.encoder import GPTEncoder
+from finetune.base_models.textcnn.featurizer import textcnn_featurizer
+
+class TextCNNModel(SourceModel):
+
+    encoder = GPTEncoder
+    featurizer = textcnn_featurizer
+    kernel_sizes = [2, 3, 4]
+    settings = {
+        'n_embed_featurizer': 768,
```
Comment (benleetownsend) [inline finetune/base_models/textcnn/model.py]: This should be made consistent across the other models, or a comment and exceptions to show, for example, that if they are set but are different, this cant work on the other models.

File: finetune/base_models/textcnn/model.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+from finetune.base_models import SourceModel
+from finetune.base_models.gpt.encoder import GPTEncoder
+from finetune.base_models.textcnn.featurizer import textcnn_featurizer
+
+class TextCNNModel(SourceModel):
+
+    encoder = GPTEncoder
+    featurizer = textcnn_featurizer
+    kernel_sizes = [2, 3, 4]
```
Comment (benleetownsend) [inline finetune/base_models/textcnn/model.py]: Worth running some experiments on enso to find some good defaults for this.

File: finetune/base_models/textcnn/featurizer.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        conv_layers = []
+        for i, kernel_size in enumerate(config.kernel_sizes):
+            conv = tf.layers.conv1d(
+                inputs=h,
+                filters=config.num_filters_per_size,
+                kernel_size=kernel_size,
+                padding='same',
+                activation=tf.nn.relu,
+                name='conv' + str(i)
+            )
```
Comment (benleetownsend) [inline finetune/base_models/textcnn/featurizer.py]: Worth thinking about reguarisation for this part of the model if you havn't already. We have weight decay by default but it might be that more is required?

File: finetune/base_models/textcnn/model.py (hunk 5 above)
Comment (lhz1029) [reply finetune/base_models/textcnn/model.py]: Discussed offline, will deal with this in a future PR

File: finetune/base_models/textcnn/featurizer.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                name='conv' + str(i),
+                kernel_initializer=tf.initializers.glorot_normal
+            )
+            conv_layers.append(conv)
+            pool = tf.reduce_max(conv, axis=1)
+            pool_layers.append(pool)
+
+        # Concat the output of the convolutional layers for use in sequence embedding
+        conv_seq = tf.concat(conv_layers, axis=2)
+        clf_token = encoder['_classify_']
```
Comment (madisonmay) [inline finetune/base_models/textcnn/featurizer.py]: Do we want to use the state of the _classify_ token as the sequence representation?  Seems more typical to use pool sequence states for conv models w/o attn like this

File: finetune/base_models/textcnn/model.py (hunk 6 above)
Comment (madisonmay) [reply finetune/base_models/textcnn/model.py]: May also make sense to include unigram features.

File: finetune/base_models/textcnn/model.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    settings = {
+        'n_embed_featurizer': 768,
+        'n_layer': 1,
+        'num_layers_trained': 1,
+        'kernel_sizes': kernel_sizes,
+        'num_filters_per_size': 2,
```
Comment (madisonmay) [inline finetune/base_models/textcnn/model.py]: This is one where I could see sensical value for this model being anywhere from 1 - 100.  Definitely worth looking into, like that you started w/ a small value though.

File: finetune/base_models/textcnn/featurizer.py (hunk 7 above)
Comment (madisonmay) [reply finetune/base_models/textcnn/featurizer.py]: Base paper used a pretty high dropout_rate on the conv layer of 0.5

Comment (madisonmay) [reply finetune/base_models/textcnn/featurizer.py]: image

File: tests/test_association.py [hunk 10]
```diff
@@ -33,6 +33,7 @@ def setUp(self):
         self.texts, self.labels = data
         self.model = Association(batch_size=2, max_length=32, **schema)
 
+    @unittest.skip('LM')
```
Comment (madisonmay) [inline tests/test_association.py]: Checkout `@unittest.skipIf`
